concert/experiments/addons.py

```diff
@@ -203,8 +202,128 @@ def _detach(self):
             acq.consumers.remove(consumer)
 
 
-class AddonError(Exception):
+class PhaseGratingSteppingFourierProcessing(Addon):
+    """
+    Addon for concert.experiments.imaging.GratingInterferometryStepping to process the raw data.
+    The order of the acquisitions can be changed.
+    """
+
+    def __init__(self, experiment, output_directory="contrasts"):
+        from concert.experiments.imaging import GratingInterferometryStepping
+        if not isinstance(experiment, GratingInterferometryStepping):
+            raise Exception("This addon can only be used with "
+                            "concert.experiments.imaging.GratingInterferometryStepping.")
+        self._output_directory = output_directory
+        self._consumers = {}
+        self._experiment = experiment
+        self._dark_image = None
+        self._reference_stepping = []
+        self._object_stepping = []
+        super().__init__(experiment.acquisitions)
 
+    def _attach(self):
+        self._consumers[self._experiment.get_acquisition("darks")] = self.process_darks
+        self._consumers[self._experiment.get_acquisition(
+            "reference_stepping")] = self.process_reference_stepping
+        self._consumers[self._experiment.get_acquisition(
+            "object_stepping")] = self.process_object_stepping
+
+        for acq, consumer in self._consumers.items():
+            acq.consumers.append(consumer)
+
+    def _detach(self):
+        for acq, consumer in list(self._consumers.items()):
+            acq.consumers.remove(consumer)
+
+    async def process_darks(self, producer):
+        self._dark_image = None
+        async for item in producer:
+            if self._dark_image is None:
+                self._dark_image = item.astype(np.float32)
+            else:
+                self._dark_image += item
+        self._dark_image /= await self._experiment.get_num_darks()
+
+    async def process_reference_stepping(self, producer):
+        self._reference_stepping = []
+        async for item in producer:
+            self._reference_stepping.append(item)
+        if self._experiment.acquisitions[-1] == self._experiment.get_acquisition(
+                "reference_stepping"):
+            await self._compare_and_write()
+
+    async def process_object_stepping(self, producer):
+        self._object_stepping = []
+        async for item in producer:
+            self._object_stepping.append(item)
+        if self._experiment.acquisitions[-1] == self._experiment.get_acquisition("object_stepping"):
+            await self._compare_and_write()
```

Comment:
The above two methods are almost identical and can have a common parametrized ancestor instead of the repetition.

Reply:
Do you have a suggestion how to make this in a clean way?

Reply:
I have an idea how to restructure this.

Reply:
Done